### ISSUE TITLE
oci_dns_resolver deletes resource from TF state file without dropping view attachment and rules

### ISSUE THREAD

**Issue by imrannayer** [oracle/terraform-provider-oci#1240]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform Version and Provider Version

```
Terraform v0.12.21
oci v4.4.0

```

### Affected Resource(s)
`oci_dns_resolver`

### Terraform Configuration Files

```hcl
resource "oci_dns_resolver" "resolver" {
  resolver_id = data.oci_core_vcn_dns_resolver_association.vcn_dns_resolver_association.dns_resolver_id
  scope       = "PRIVATE"

  ### Attach DNS views

  attached_views {
    view_id = oci_dns_view.view.id
  }

  ### Add forwarding Rules
  rules {
    destination_addresses  = ["10.0.0.2"]
    action                 = "FORWARD"
    source_endpoint_name   = oci_dns_resolver_endpoint.resolver_endpoint.name
    qname_cover_conditions = ["gcr.io"]
  }
}
```

### Expected Behavior

If I comment out resource `oci_dns_resolver` it should remove attached views and rules from the resolver before deleteing resolver from tfstate file. Here is an example code:

```
#   resource "oci_dns_resolver" "resolver" {
#   resolver_id = data.oci_core_vcn_dns_resolver_association.vcn_dns_resolver_association.dns_resolver_id
#   scope       = "PRIVATE"

#   ### Attach DNS views

#   attached_views {
#     view_id = oci_dns_view.view.id
#   }

#   ### Add forwarding Rules
#   rules {
#     destination_addresses  = ["10.0.0.2"]
#     action                 = "FORWARD"
#     source_endpoint_name   = oci_dns_resolver_endpoint.resolver_endpoint.name
#     qname_cover_conditions = ["gcr.io"]
#   }
# }

```
### Actual Behavior
Instead if just removes the resolver from state file and leaves attached rules/DNS views in place. 

### Steps to Reproduce

1. Update VCN resolver with attached views and forwarding rules using `oci_dns_resolver` resource

```
resource "oci_dns_resolver" "resolver" {
  resolver_id = data.oci_core_vcn_dns_resolver_association.vcn_dns_resolver_association.dns_resolver_id
  scope       = "PRIVATE"

  ### Attach DNS views

  attached_views {
    view_id = oci_dns_view.view.id
  }

  ### Add forwarding Rules
  rules {
    destination_addresses  = ["10.0.0.2"]
    action                 = "FORWARD"
    source_endpoint_name   = oci_dns_resolver_endpoint.resolver_endpoint.name
    qname_cover_conditions = ["gcr.io"]
  }
}

```

2. Comment out the code and execute `terraform apply`. You can still see views attached and forwarding rules in resolver. 

```
#   resource "oci_dns_resolver" "resolver" {
#   resolver_id = data.oci_core_vcn_dns_resolver_association.vcn_dns_resolver_association.dns_resolver_id
#   scope       = "PRIVATE"

#   ### Attach DNS views

#   attached_views {
#     view_id = oci_dns_view.view.id
#   }

#   ### Add forwarding Rules
#   rules {
#     destination_addresses  = ["10.0.0.2"]
#     action                 = "FORWARD"
#     source_endpoint_name   = oci_dns_resolver_endpoint.resolver_endpoint.name
#     qname_cover_conditions = ["gcr.io"]
#   }
# }

```
3. Instead of commenting out whol resouce `oci_dns_resolver` if you just comment out attached_views or rules it works fine. But this will make it a 2 step process for removing DNS settings. Here is an example code which works fine. 

```
resource "oci_dns_resolver" "resolver" {
  resolver_id = data.oci_core_vcn_dns_resolver_association.vcn_dns_resolver_association.dns_resolver_id
  scope       = "PRIVATE"

  ### Attach DNS views

  # attached_views {
  #   view_id = oci_dns_view.view.id
  # }

  # ### Add forwarding Rules
  # rules {
  #   destination_addresses  = ["10.0.0.2"]
  #   action                 = "FORWARD"
  #   source_endpoint_name   = oci_dns_resolver_endpoint.resolver_endpoint.name
  #   qname_cover_conditions = ["gcr.io"]
  # }
}

```

**Comment by zexinwanoci** [comment]
Hello, imrannayer, unfortunately since terraform do not manage the creation or deletion of the dns resolver, this is the expected behavior. You can find this in the document
`Note: Resolvers are associated with VCNs and created when a VCN is created. Wait until created VCN's state shows as Available in OCI console before updating DNS resolver properties.`

If you do not like this behavior, let us know and we can redirect you to the networking service team.

**Comment by imrannayer** [comment]
I know `oci_dns_resolver` does not create/destroy VCN resolver. It is used to attach DNS view or forwarding rules to the VCN resolver. 
The problem is when we remove `oci_dns_resolver` code it should detach views from resolver and remove forwarding rules before removing `oci_dns_resolver` from tfstate file. Example code is in  step 2 of "Steps to Reproduce"
Current behavior is a bug as it works fine if I just remove rules or attached views block and keep `oci_dns_resolver` in code. Example code is in step 3 of "Steps to Reproduce"

**Comment by zexinwanoci** [comment]
Hello imrannayer, to remove views and rules attached to VCN resolver, you need to do what you did for step3, which is an update request that we support. The only way to delete `oci_dns_resolver` is to delete the attached VCN. Commenting out `oci_dns_resolver` will not remove the resource from backend is a design limitation and it is expected to behave like this. If you just want to remove the rules and attached views then what you did in step 3 is good enough. You don't need to comment out the `oci_dns_resolver` resource. If you do not like how it is handled right now, please let us know and we can redirect you to the service team.

**Comment by imrannayer** [comment]
@zexinwanoci I can comment out rules and attached views as mentioned in step 3 but it will be a problem when I execute terraform destroy. 
TF will leave views and rules attached to resolver due to which I will not be able to delete vcn. You can reproduce this issue by simply creating the following resources using terraform and then execute terraform destroy. 

Create TF script with following resources

- Create VCN
- Create the view
-  attach the dns view to vcn resolver

Execute terraform apply

Once resources are created execute terraform destroy. It will fail. It will fail to delete vcn cos `oci_dns_resolver` will not detach views.

**Comment by rashik-bhasin** [comment]
Hi @imrannayer, 
I tried the example where i had attached views in my resolver and I was able to delete the VCN. Can you please try again, we have an example explaining this behavior as well here: https://github.com/terraform-providers/terraform-provider-oci/tree/master/examples/dns/private/resolver . 
Secondly, Resolver endpoints can't be deleted when they are used in one of the resolver's rules though. So if a user has a vcn/resolver/resolver endpoint and the resolver has a rule pointing to the resolver endpoint then in order to delete the vcn:
1). You will have to update the resolver and remove rule from resolver 
2). And delete resolver endpoint.
3). then you will be able to delete vcn.

So in conclusion, a VCN can be deleted while the resolver has attachedViews and you might be getting blocked on vcn deletion from the presence of a resolver endpoint.

Thanks!

**Comment by imrannayer** [comment]
@rashik-bhasin I dont think my issue is about resolver as resolver is part of VCN and created/deleted as part of VCN. I fully understand the flow. 
Bug I reported here is that when I execute terraform destroy provider is deleting `oci_dns_resolver` resource in tfstate file without deleting the rules or detatching views to it which results in failure to delete vcn in same script. Due to this bug deleting VCN is 2 step process:
1) Comment lines with rules and attached views execute terraform apply. 
2) Execute terraform destroy. 

When the provider deletes the` oci_dns_resolver` resource (I know it does not delete the resolver just removes it from the state file) it should delete rules and detach views. 

You can test it by commenting out `oci_dns_resolver` resource in your code and execute terraform apply. Once code is executed go to console and check resolver. You will find rules and views still attached to resolver.

**Comment by tjzepf** [comment]
@imrannayer is correct - this is a bug and a truly awful one. A resource should always clean up after itself on a destroy, and in this case it does not. Yes, the resolver cannot be deleted until the VCN is destroyed - no problem. But any rules created by this resource *should* be cleaned up on destroy.

This bug prevents any terraform code using this resource from destroying. Please bump the priority of this bug up and resolve it.

**Comment by tzepf** [comment]
@rashik-bhasin why was the bug label removed. This is truly a bug.

**Comment by ITD27M01** [comment]
I can confirm that it is a big problem for TF automation, any rules and attachments described in the TF resolver resource should be deleted from resolver before deleting it from TF state.

**Comment by ITD27M01** [comment]
In general, the root cause of such problems is a practice to manage external dependent objects inside one API update request.

For this issue case the rules and attachments could be defined in separated resources and they would be deleted during _terraform destroy_ as terraform graph dependencies:

```terraform
resource "oci_dns_resolver" "resolver" {
  resolver_id = data.oci_core_vcn_dns_resolver_association.vcn_dns_resolver_association.dns_resolver_id
  scope       = "PRIVATE"
}

resource "oci_dns_resolver_rule" "resolver_rule" {
    count = lenght(var.forwarding_rules)

    resolver_id = oci_dns_resolver.resolver.id // link to above resource

    destination_addresses  = var.forwarding_rules.addresses
    action                 = "FORWARD"
    source_endpoint_name   = oci_dns_resolver_endpoint.resolver_endpoint.name
    qname_cover_conditions = var.forwarding_rules.domains
}

resource "oci_dns_resolver_attachment" "resolver_attachment" {
    resolver_id = oci_dns_resolver.resolver.id // link to above resource
    view_id = oci_dns_view.view.id
}
```

The issue #1251 has the same nature. 

But if we dig deeper, the reason for such limitations is premature designed API. The attached views and rules all weren't decorated as separated objects but just fields inside UpdateResolverDetails:
https://docs.oracle.com/en-us/iaas/api/#/en/dns/20180115/datatypes/UpdateResolverDetails

There are a lot of examples for other OCI services with the same problem.

So, as for now, it is difficult to deal with it on the client side (terraform oci provider) and automation for Oracle Cloud will be full of manual interventions that OCI engineers will call "workarounds".

**Comment by ocimatix-pw** [comment]
Hi community - FWIW, we are also struggling with that issue (and others of similar nature) in our automation framework, we have created also an SR for that matter: SR 3-27743984281 : DNS rules prevent terraform destroy from running
Will keep you posted here about any updates for this.
- Peter

**Comment by ocimatix-pw** [comment]
FWIW, quick update on above SR - in the meantime it is on Oracle Dev backlog with distinct Jira ticket. However, we did not get an ETA yet. Will keep you posted here i.e. Oracle wanted to post here once the issue is solved (fix delivered).

**Comment by vsin12** [comment]
Fix for the issue has been released with terraform version https://registry.terraform.io/providers/oracle/oci/latest
